NOAA-OWP/ras2fim
# [1pt] PR: Download OWP Models by HUC

Formally called get_ras2fim_by_catalog, this program allows for a user to provide an S3 path and a HUC, have it query the remote models catalog csv and download "unprocessed" folders to their local machine for processing with ras2fim.py.

Key Notes
- `This tool is designed for working with OWP S3 and will likely not have value to non NOAA staff`, unless they have their own AWS account with a similar structure.  However, ras2fim.py will continue to work without the use of this tool.
- Review the arguments being passed into `get_ras_models.py` file which include defining your own `models catalog csv` file but it will default to `models_catalog.csv`.  ie). testing, it could be `models_catalog_rob.csv`.
- New dependency python packages were added. 

Associated to issue [39](https://github.com/NOAA-OWP/ras2fim/issues/39). 

### Additions  
- `tools/aws`
    - `aws_base.py`: a basic parent class used to provide common functionality to any aws type code calls. Currently, it is used by `get_ras_models.py`, but others are expected.
    - `get_ras_models.py` : The tool mentioned above. See key features below.
    - `aws_creds_templates.env`: At this point, some functionality requires a valid aws credentials. However, a user will still need to setup their local user .aws folder with a config and credentials file in it, for this program to run. These are hoped to be amalgamated later. One of the args into `get_ras_models.py` is an override for the default location and adjusted file name.

### Changes  
- `doc\INSTALL.md`: minor corrections

### Features for this tool include:
-  Ability to create a list only (log file) without actual downloads (test download).
- For both pulling from S3 as a source as well as for the local target, the "models" will automatically be added at the end of thos  provided arguments if/as required.
- A log file be created in the models/log folder with unique date stamps per run.
- A "verbose" flag can be optionally added as an argument for additional processing details (note: don't over  use this as it can make errors harder to find). To find an error, simply search the log for the word "error".
- Filters downloads from the src models catalog to look for status of "ready" only. Also filters out model catalog final_key_names starting with "1_" one underscore.
- Can find huc numbers in the models catalog "hucs" field regardless of string format in that column.  It does assume that models catalog has ensure that leading zero's exist.
 - All folders in the target models folder will be removed at the start of execution unless the folder starts with two underscores. This ensures old data from previous runs does not bleed through. 
- After the source model catalog has been filtered, it will save a copy called models_catalog_{YYYYMMDD}.csv into the target output folder.  Later this csv can be feed into ras2fim.py for meta data for each model.

### Steps to apply new features including AWS credentials and conda enviroment.yml changes.

While this is a number of steps, it also setups for a fair bit of new and future functionality.

Use your anaconda terminal window for all "type" commands below.

1) Path to your new code dir. ie) cd c:\users\(your profile name}\projects\dev\ras2fim.
2) Type `conda activate ras2fim`
3) Type `conda env update -f environment.yml`.  You are ready to do testing with `ras2fim.py` and other files, but not `get_ras_models.py`

**If you are an NOAA/OWP staff member, do the next steps. Only NOAA/OWP members can use the new `get_ras_models.py`.**
1) See if you already have aws credentials file in place. Path to c:\users\{profile name|\.aws with a file named config and another named credentials in it. You might have to turn on with file explorer, the ability to see hidden files / folders. Once this is setup for your machine, it will not need to be done again.
2) If that folder and files are not there, type `aws configure`. It will ask for some AWS keys which you should already have.
3) In the c:\ras2fim_data folder, create a new subfolder called "config"
4) Inside the code folder "\tools\aws" is a file called aws_cred_template.env. Copy that file to your new "config" folder.
5) Rename that file to "aws_hv_s3_creds.env"
6) Open the file and fill in the fields with the aws credentials values you should have. Make sure to save it.
7) You are ready to test the new tool.

### Testing

A wide range of tests were performed including:
- various combinations against the models catalog file, hucs field, including valid HUCs, invalid HUCs, hucs with no row, rows with no hucs, invalid huc formats, etc.
- Various error conditions were triggered to ensure all exceptions were caught elegantly and logged when desired. ie) bad args errors are not logged.
- Various scenarios of target folders pre-existing or not existing including skipping removing folders starting with two underscores.

### How to test the new tool (OWP staff only)
1) By default, this will remove all folders in the C:\ras2fim_data\OWP_ras_models\models, unless the subfolders start with the two  underscores. So.. check what you have in that folder before you start. This script will not remove files.
2) In the anaconda terminal window with ras2fim activated, type: 
python .\tools\aws\get_ras_models.py -s s3://ras2fim/OWP_ras_models_robh/model_catalog_robh.csv -u 08010202. 
You should see three folders in the C:\ras2fim_data\OWP_ras_models\models, PLUS a log file in C:\ras2fim_data\OWP_ras_models\logs, PLUS a filtered copy of the models catalog file in s3 that was hit in the first place. ie) we bounced against s3://ras2fim/OWP_ras_models_robh/model_catalog_robh.csv, so you should see a filtered copy with a date stamp attached. Note: The records in there are all that were attempted to be downloaded. Some may have been skipped during download so watch the log files.
3) Read the documentation about the parameters and how the script works.


A volume tests were not performed, but appears to be running at appx 2 to 5 seconds per model folder to be downloaded.
### Checklist

_You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code._

- [x] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [x] Links are provided if this PR resolves an issue, or depends on another other PR
- [x] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g.: `dev-revise-levee-masking`)
- [x] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [x] Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
- [ ] Passes all unit tests locally (inside interactive Docker container, at `/foss_fim/`, run: `pytest unit_tests/`)
- [ ] Project documentation has been updated ([CHANGELOG](/docs/CHANGELOG.md) and/or [README](/README.md))
- [ ] [Reviewers requested](https://help.github.com/articles/requesting-a-pull-request-review/)
